File: CodeBattleJava/src/main/java/ru/codebattle/client/api/GameBoard.java
package ru.codebattle.client.api;

import static ru.codebattle.client.api.BoardElement.APPLE;
import static ru.codebattle.client.api.BoardElement.ENEMY_HEAD_SLEEP;
import static ru.codebattle.client.api.BoardElement.ENEMY_TAIL_INACTIVE;
import static ru.codebattle.client.api.BoardElement.FLYING_PILL;
import static ru.codebattle.client.api.BoardElement.FURY_PILL;
import static ru.codebattle.client.api.BoardElement.GOLD;
import static ru.codebattle.client.api.BoardElement.HEAD_DEAD;
import static ru.codebattle.client.api.BoardElement.HEAD_DOWN;
import static ru.codebattle.client.api.BoardElement.HEAD_EVIL;
import static ru.codebattle.client.api.BoardElement.HEAD_FLY;
import static ru.codebattle.client.api.BoardElement.HEAD_LEFT;
import static ru.codebattle.client.api.BoardElement.HEAD_RIGHT;
import static ru.codebattle.client.api.BoardElement.HEAD_SLEEP;
import static ru.codebattle.client.api.BoardElement.HEAD_UP;
import static ru.codebattle.client.api.BoardElement.START_FLOOR;
import static ru.codebattle.client.api.BoardElement.STONE;
import static ru.codebattle.client.api.BoardElement.TAIL_INACTIVE;
import static ru.codebattle.client.api.BoardElement.WALL;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import lombok.Getter;

public class GameBoard {

  public GameBoard(String boardString) {
    this.boardString = boardString.replace("\n", "");
  }

  @Getter
  private String boardString;

  public int size() {
    return (int) Math.sqrt(boardString.length());
  }

  public BoardPoint getMyHead() {
      return findFirstElement(HEAD_DEAD, HEAD_DOWN, HEAD_UP, HEAD_LEFT, HEAD_RIGHT, HEAD_EVIL,
          HEAD_FLY, HEAD_SLEEP);
  }

  public List<BoardPoint> getWalls() {
      return findAllElements(WALL);
  }

  public List<BoardPoint> getStones() {
      return findAllElements(STONE);
  }

  public boolean isBarrierAt(BoardPoint point) {
      return getBarriers().contains(point);
  }

  public List<BoardPoint> getApples() {
      return findAllElements(APPLE);
  }

  public boolean amIEvil() {
      return findAllElements(HEAD_EVIL).contains(getMyHead());
  }

  public boolean amIFlying() {
      return findAllElements(HEAD_FLY).contains(getMyHead());
  }

  public List<BoardPoint> getFlyingPills() {
      return findAllElements(FLYING_PILL);
  }

  public List<BoardPoint> getFuryPills() {
      return findAllElements(FURY_PILL);
  }

  public List<BoardPoint> getGold() {
      return findAllElements(GOLD);
  }

  public List<BoardPoint> getStartPoints() {
      return findAllElements(START_FLOOR);
  }

  private List<BoardPoint> getBarriers() {
      return findAllElements(WALL, START_FLOOR, ENEMY_HEAD_SLEEP, ENEMY_TAIL_INACTIVE, TAIL_INACTIVE, STONE);
  }

  public boolean hasElementAt(BoardPoint point, BoardElement element) {
    if (point.isOutOfBoard(size())) {
      return false;
    }

    return getElementAt(point) == element;
  }

  public BoardElement getElementAt(BoardPoint point) {
    return BoardElement.valueOf(boardString.charAt(getShiftByPoint(point)));
  }

  public void printBoard() {
    for (int i = 0; i < size(); i++) {
      System.out.println(boardString.substring(i * size(), size() * (i + 1)));
    }
  }

  public BoardPoint findElement(BoardElement elementType) {
    for (int i = 0; i < size() * size(); i++) {
      BoardPoint pt = getPointByShift(i);
      if (hasElementAt(pt, elementType)) {
        return pt;
      }
    }
    return null;
  }

  public BoardPoint findFirstElement(BoardElement... elementType) {
      for (int i = 0; i < size() * size(); i++) {
          BoardPoint pt = getPointByShift(i);

          for (BoardElement elemType : elementType) {
              if (hasElementAt(pt, elemType)) {
                  return pt;
              }
          }
      }
      return null;
  }

  public List<BoardPoint> findAllElements(BoardElement... elementType) {
    List<BoardPoint> result = new ArrayList<>();

    for (int i = 0; i < size() * size(); i++) {
      BoardPoint pt = getPointByShift(i);

      for (BoardElement elemType : elementType) {
        if (hasElementAt(pt, elemType)) {
          result.add(pt);
        }
      }
    }

    return result;
  }

  public boolean hasElementAt(BoardPoint point, BoardElement... elements) {
    return Arrays.stream(elements).anyMatch(element -> hasElementAt(point, element));
  }

  private int getShiftByPoint(BoardPoint point) {
    return point.getY() * size() + point.getX();
  }

  private BoardPoint getPointByShift(int shift) {
    return new BoardPoint(shift % size(), shift / size());
  }
}

File: CodeBattleCPP/CodeBattleCppLibrary/GameBoard.cpp
#include "GameBoard.h"

GameBoard::GameBoard(BoardElement** map, int map_size)
{
	this->map = map;
	this->map_size = map_size;
}


bool GameBoard::hasElementAt(BoardPoint point, BoardElement element) {
	if (point.isOutOfBoard(map_size)) {
		return false;
	}
	return getElementAt(point) == element;
}
BoardElement GameBoard::getElementAt(BoardPoint point) {
	return map[point.getX()][point.getY()];
}
std::list<BoardPoint> GameBoard::findAllElements(BoardElement element) {
	std::list<BoardPoint> result;
	for (uint32_t j = 0; j < map_size; j++)
	{
		for (uint32_t i = 0; i < map_size; i++)
		{
			if (map[j][i] == element) {
				result.push_back(BoardPoint(j, i));
			}
		}
	}
	return result;
}
void GameBoard::printBoard() {

}
BoardPoint GameBoard::getMyHead() {
	std::list<BoardPoint> result = findAllElements(BoardElement::HEAD_DEAD);
	result.splice(result.end(), findAllElements(BoardElement::HEAD_DOWN));
	result.splice(result.end(), findAllElements(BoardElement::HEAD_UP));
	result.splice(result.end(), findAllElements(BoardElement::HEAD_LEFT));
	result.splice(result.end(), findAllElements(BoardElement::HEAD_RIGHT));
	result.splice(result.end(), findAllElements(BoardElement::HEAD_EVIL));
	result.splice(result.end(), findAllElements(BoardElement::HEAD_FLY));
	result.splice(result.end(), findAllElements(BoardElement::HEAD_SLEEP));
	return result.front();
}
std::list<BoardPoint> GameBoard::getWalls() {
	return findAllElements(BoardElement::WALL);
}
bool GameBoard::isBarrierAt(BoardPoint point) {
	std::list<BoardPoint> result = getBarriers();
	BoardPoint resultPoint = result.front();
	return resultPoint == point;
}
std::list<BoardPoint> GameBoard::getStones() {
	return findAllElements(BoardElement::STONE);
}
std::list<BoardPoint> GameBoard::getApples() {
	return findAllElements(BoardElement::APPLE);
}
bool GameBoard::amIEvil() {
	std::list<BoardPoint> headEvil = findAllElements(BoardElement::HEAD_EVIL);
	if (!headEvil.empty()) {
		return true;
	}
	return false;
}
bool GameBoard::amIFlying() {
	std::list<BoardPoint> headFly = findAllElements(BoardElement::HEAD_FLY);
	if (!headFly.empty()) {
		return true;
	}
	return false;
}
std::list<BoardPoint> GameBoard::getFlyingPills() {
	return findAllElements(BoardElement::FLYING_PILL);
}
std::list<BoardPoint> GameBoard::getFuryPills() {
	return findAllElements(BoardElement::FURY_PILL);
}
std::list<BoardPoint> GameBoard::getGold() {
	return findAllElements(BoardElement::GOLD);
}
std::list<BoardPoint> GameBoard::getStartPoints() {
	return findAllElements(BoardElement::START_FLOOR);
}
std::list<BoardPoint> GameBoard::getBarriers() {
	std::list<BoardPoint> result = findAllElements(BoardElement::WALL);
	result.splice(result.end(), findAllElements(BoardElement::START_FLOOR));
	result.splice(result.end(), findAllElements(BoardElement::ENEMY_HEAD_SLEEP));
	result.splice(result.end(), findAllElements(BoardElement::ENEMY_TAIL_INACTIVE));
	result.splice(result.end(), findAllElements(BoardElement::TAIL_INACTIVE));
	result.splice(result.end(), findAllElements(BoardElement::STONE));
	return result;
}

GameBoard::~GameBoard()
{
	delete map;
}

